### ISSUE TITLE
Dagger unused member injection check

### ISSUE THREAD

**Issue by JakeWharton** [google/error-prone#593]
When doing members injection with Dagger, IDEs can no longer detect fields as unused due to the members injector referencing every field. Having error-prone fill this gap can ensure the minimal subset of bindings are always used when doing members injection.

Example:
```java
class FooActivity extends Activity {
  @Inject BarManager barManager;
  @Inject BazManager bazManager;

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    TextView tv = new TextView(this);
    tv.setText(bazManager.getBaz());
   setContentView(tv);
  }
}
```

**Comment by stephanenicolas** [comment]
Though it becomes a bit more tricky than private fields because fields can used by subclasses (or even other classes in the same package, that would be dirty but it's allowed). It would be nice to have but how can a rule understand that some fields are parts of the API exposed to subclasses?

Or we would detect these as unused an annotate them to ignore them?

**Comment by ronshapiro** [comment]
It's probably a bad practice to inject something in a base class and never use it just so that a subclass can.  How do you know the subclass will need it? Shouldn't you just let them inject it?

**Comment by stephanenicolas** [comment]
In an abstract class for instance.

2017-10-08 19:00 GMT-07:00 Ron Shapiro <notifications@github.com>:

**Comment by JakeWharton** [comment]
It would either be used in the same compilation unit and not error,
protected and not error, or the error would be suppressed.

**Comment by stephanenicolas** [comment]
This makes sense. It's not a good practice but a field could be public too,
and not error.
(I checked the error prone code base quickly and couldn't find any similar
rule to get started. Pmd's unused field rule would be easy to adapt.)

2017-10-08 20:31 GMT-07:00 Jake Wharton <notifications@github.com>:
